presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java

```diff
@@ -146,7 +147,8 @@ public PlanOptimizers(
             StatsCalculator statsCalculator,
             CostCalculator costCalculator,
             @EstimatedExchanges CostCalculator estimatedExchangesCostCalculator,
-            CostComparator costComparator)
+            CostComparator costComparator,
+            TaskCountEstimator taskCountEstimator)
```

Comment:
I don't like this extra dependency here. This class is becoming a grab bag of everything someone might ever need.

Why is taskCountEstimator not an implementation detail of CostCalculator?

Reply:
`CostCalculator` is an interface that has 2 different implementation. It hash a single method, `calculateCost`, that computes cost of the entire operation. For joins this is the cost of exchanges needed for Join, cost of the Join itself, and cost of producing output out of Join. If the join cardinality cannot be estimated - the total cost of join cannot be estimated as well. What we are trying to achive here, is to be able to estimate cost of Join exchanges + cost of Join input processing. So even if the total join cost cannot be estimated, the sides still can be swapped.

Reply:
>  2 different implementation

One implementation works when exchanges were already introduced. The other one assumes there no exchange nodes introduced, so it adds the estimated cost of exchanges.

Reply:
As an alternative, i can add aditional non static methods specifically to the `CostCalculatorWithEstimatedExchanges`, and use that class directly (not by the `CostCalculator` interface) in the `DetermineJoinDistributionType`.

@martint What do you thin?

CC: @rschlussel @findepi @sopel39

Reply:
As i noted in https://github.com/prestodb/presto/pull/11982#discussion_r237281110:
> Cost calculators are passed to Rules from IterativeOptimizer, via the Context

if you want to change this design, we can talk about this. However, if this stays, then i think it should remain `IterativeOptimizer`'s responsibility to provide contextual information that rules need.

OTOH, would it work to implement `DetermineJoinDistributionType` in a simpler manner, based on stats calculator alone and not considering contextual information like number of nodes? I know considering more is supposed to yield better plans, but here we're in "rescue mode" anyway. Basing plan on stats alone (bypassing cost calculator) is not the route we want to take frequently.
@arhimondr would that work?

Reply:
> would it work to implement DetermineJoinDistributionType in a simpler manner, based on stats calculator alone and not considering contextual information like number of nodes?

It is easy to implement join sides flipping based on stats. However it is not that easy to implement replicated join selection based on stats. As replicated over repartitioned join requires different resources (more memory, more network, less cpu), and it is important to apply the cost weights there.

> but here we're in "rescue mode" anyway

For us this is gonna be a very common situation. Unfortunately we don't have NDV's for many tables. And without NDV it is not possible to estimate the JOIN output.

Reply:
Could you use `CostCalculatorUsingExchanges` calculator to calculate cost of join source plans with exchanges (created for the purpose of comparison) added? I think you did something similar for `UnionNode`.

Then you won't need to use directly `TaskCountEstimator`?

Reply:
> Could you use CostCalculatorUsingExchanges calculator to calculate cost of join source plans with synthetic exchange

Based on the offline discussion we agreed that it is not possible. As CostCalculatorUsingExchanges estimates the total cost of a JOIN, which cannot be estimated if the output JOIN cardinality cannot be estimated.